datasets/cfq/data_generation.py

```diff
@@ -0,0 +1,341 @@
+        random: Optional; boolean indicating whether to permute the remaining 
+          clauses (which haven't been ordered according to word order and csv_map)
+          at random.
+
+    Returns:
+        A modified SPARQL query ordered according to the word order in question
+          and csv_map.
+
+    """
+    words = question.split(" ")
+    clauses = query.split("\n")
+    prefix = clauses[0]
+    postfix = clauses[-1]
+    clauses = clauses[1:-1]
+    
+    if random:
+      rd.shuffle(clauses)
+      
+    if csv_map is not None:
+      ranking_list = []
+      for i, word in enumerate(words):
+        if word in csv_map.keys():
+            score = len(words) - i
+            this_ranking = np.zeros(len(clauses))
+            for j, clause in enumerate(clauses):
+                for value in csv_map[word]:
+                    if value + " " in clause and this_ranking[j] == 0:
+                        this_ranking[j] = score
+            ranking_list.append(this_ranking)
+    
+      ranking_array = np.array(ranking_list)
+      clause_ordering = np.argsort(-1 * np.sum(ranking_array, axis=0))
+      clauses = [clauses[i] for i in clause_ordering] 
+    
+    for idx, clause in enumerate(clauses):
+      if idx == len(clauses) - 1:
+        if " ." in clause:
+            clauses[idx] = clause.replace(" .", "")  
+      elif " ." not in clause:
+        clauses[idx] += " ."
+    out_list = [prefix] + clauses + [postfix]
+    return "\n".join(out_list)
+      
+
+def generate_seq2seq_examples(ds, filename_in, filename_out, simplify_output=False,
+                              permute=False, respect_question_order=False, random=False):
+    input_file = open(filename_in, "w")
+    output_file = open(filename_out, "w")
+    
+    csv_map = None
+    if respect_question_order:
+      csv_map = dict(dict_of_args)
+      with open("cfq_map.csv", mode="r") as csv_file:
+        csv_reader = csv.reader(csv_file, delimiter=",")
+        for row in csv_reader:
+            if row[0] not in csv_map.keys():
+                csv_map[str(row[0])] = [str(row[1])]
+            else:
+                csv_map[str(row[0])].append(str(row[1]))
+    
+    for example in ds: 
+      query = example["query"].numpy().decode("utf-8").strip()
+      query = query.replace("\n", " ")
+      if simplify_output:
+        query = simplify_cfq_output(query)
+      question = example["question"].numpy().decode("utf-8").strip()
+      if permute:
+        query = query.replace("{ ", "{\n")
+        query = query.replace(". ", ".\n")
+        query = query.replace(" }", "\n}")
+        query = permute_clauses(question, query, csv_map, random)
+        query = query.replace("\n", " ")
+      
+      input_file.write(question.strip("\n") + "\n")
+      output_file.write(query.strip("\n") + "\n")
+    
+    input_file.close()
+    output_file.close()
+
+
+def generate_it_dec_examples(ds, split, filename_in, filename_out, filename_ops=None, 
```

Comment:
The rest of the file has a mixture of 2-space and 4-space indentation.  The Google style guide recommends 2 spaces throughout.